### DISCUSSION TITLE
My yolo model does not work well on object detection in live video stream. Even if it has achieved great performance when training.

### DISCUSSION THREAD

**Discussion by CharliePip** [ultralytics/yolov5#2787] [Q&A] [answered]
My dear friends, I urgently need your help on Yolo. Please help.

I need to use the Yolo model to detect dumbbells in the hands of exercisers. I randomly divided 8000 labelled data into training set and validation set according to the ratio of 7:3.

My model performed well after training for 30 epochs. This model can achieve 95% accuracy and 92% recall rate.

But when I deployed the model on the server, I found that his performance was not good. He often detects the exerciser's clothes as dumbbells or fails to detect dumbbells.
![results](https://user-images.githubusercontent.com/81152190/114665153-3ef1f480-9d2f-11eb-9581-7bae103c452f.png)

Why does my model perform well in the validation set, but it does not work properly?

I have done the following measures to improve my model, but the effect is not good:
1. Considering that exercisers always use dumbbells in a fixed place in the gym to exercise, I used cameras in other areas and my mobile phone to record some videos of dumbbells, and then converted the videos to pictures and labelled the dumbbells in them. For training models.
The purpose of this is to increase the complexity of the data to prevent over-fitting

2. I randomly divide the labelled images into training set and validation set to avoid too single data caused by manual screening.

3. I use Yolo's own data enhancement function to optimize my training pictures, such as rotation.

Is it because the lower video bit rate causes poor video quality, which makes the model unable to detect dumbbells?

Can the performance be improved if the video stream is denoised or the image quality is improved, and then the model is used for detection?

At present, I hope that the effect of the model can be greatly improved. What should I do?

**Comment by glenn-jocher** [comment, marked as answer]
@CharliePip your training pipeline probably suffers from both problems below, as randomly dividing video frames between train and val sets will produce artificially high mAP and poor generalization in the real world. 

<img width="1296" alt="Screenshot 2021-04-14 at 20 04 48" src="https://user-images.githubusercontent.com/26833433/114758003-c86beb80-9d5c-11eb-81fc-62145265a9ba.png">

Here are our main suggestions:

👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users first train with all default settings before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by nolby003** [comment]
I still feel as though YOLO is a failed CNN

**Comment by nolby003** [comment]
I have classifed 5 classes with each 1000 images or more and have labels 100s per 5-7 states per class with annotations (boring to commit to) and trained using your v8, however, the limitations of your hub app prevents me from securing a solid detection for my 26 total classes, it think everything is either nothing or something else entirely and has a bounding box of one class constantly sitting in my camera viewing window on my phone, despite any accuracy from the epochs, this CNN can only successfully detect images using images it has already trained on, anything else, and it has no clue, its not smart at all, I could teach a dog more tricks.
